experimental/algorithm/LAGraph_CFL_reachability.c

```diff
@@ -0,0 +1,324 @@
+// LAGraph_CFL_reachability.c: Context-Free Language Reachability Matrix-Based Algorithm
+//------------------------------------------------------------------------------
+//
+// LAGraph, (c) 2019-2024 by The LAGraph Contributors, All Rights Reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+
+// Contributed by Ilhom Kombaev, Semyon Grigoriev, St. Petersburg State University.
+
+//------------------------------------------------------------------------------
+
+// Code is based on the "A matrix-based CFPQ algorithm" described in the following paper:
+//  * Rustam Azimov, Semyon Grigorev, "Context-Free Path Querying Using Linear Algebra"
+//  * URL: https://disser.spbu.ru/files/2022/disser_azimov.pdf
+
+#include "LG_internal.h"
+#include <LAGraphX.h>
+
+#define ERROR_RULE(msg)                                                                  \
+    {                                                                                    \
+        LG_ASSERT_MSGF(false, GrB_INVALID_VALUE, "Rule with index %ld is invalid. " msg, \
+                       i);                                                               \
+    }
+
+#define ADD_TO_MSG(...)                                                                  \
+    {                                                                                    \
+        if (msg_len == 0) {                                                              \
+            msg_len +=                                                                   \
+                snprintf(msg, LAGRAPH_MSG_LEN,                                           \
+                         "LAGraph failure (file %s, line %d): ", __FILE__, __LINE__);    \
+        }                                                                                \
+        if (msg_len < LAGRAPH_MSG_LEN) {                                                 \
+            msg_len += snprintf(msg + msg_len, LAGRAPH_MSG_LEN - msg_len, __VA_ARGS__);  \
+        }                                                                                \
+    }
+
+#define ADD_INDEX_TO_ERROR_RULE(rule, i)                                                 \
+    {                                                                                    \
+        rule.len_indexes_str += snprintf(rule.indexes_str + rule.len_indexes_str,        \
+                                         LAGRAPH_MSG_LEN - rule.len_indexes_str,         \
+                                         rule.count == 0 ? "%ld" : ", %ld", i);          \
+        rule.count++;                                                                    \
+    }
+
+#define LG_FREE_WORK                                                                     \
+    do {                                                                                 \
+        free(nnz);                                                                       \
+    } while (0)
```

Comment:
```suggestion
#define LG_FREE_WORK                                                                     \
    {                                                                                    \
        free(nnz);                                                                       \
    }
```

I prefer macros with _do-while_, but I haven't seen this design in LAGraph